doc: Add api curl examples (#28)
TheThingsIndustries/lorawan-stack-docs: doc/content/getting-started/api/_index.md

---
title: "Using the API"
description: ""

> weight: 

Author: will take a look at the whole section in another pr https://github.com/TheThingsIndustries/lorawan-stack-docs/issues/73

<!--more-->

A complete list of API endpoints is available in the [API Reference]({{< ref "reference/api" >}}). There, you can also find detailed information about [Authentication]({{< ref "reference/api/authentication" >}}) and [Field Masks]({{< ref "reference/api/field-mask" >}}).

> If you're having trouble with the HTTP API, you can always inspect requests in the Console using your browser's inspector. All of the data displayed in the Console is pulled using HTTP API requests, and this should give you some insight in to how they are formed.

## HTTP Queries

Additional fields may be specified in HTTP requests by appending them as query string parameters. For example, to request the `name`, `description`, and `locations` of devices in an `EndDeviceRegistry.Get` request, add these fields to the `field_mask` field. To get this data for device `dev1` in application `app1`:

```bash
$ curl --location --header "Authorization: Bearer NNSXS.XXXXXXXXX" --header 'Accept: application/json' \

> https://thethings.example.com/api/v3/applications/app1/devices/dev1?field_mask=name,description,locations

Reviewer: Probably nicer to use newlines for each argument, for readability.

```

The same request in tenant `tenant1` on a multi-tenant deployment:

```bash
$ curl --location --header "Authorization: Bearer NNSXS.XXXXXXXXX" --header 'Accept: application/json' \
https://tenant1.thethings.example.com/api/v3/applications/app1/devices/dev1?field_mask=name,description,locations
```

Fields may also be specified as a JSON object in a POST request.

To get a stream of events for device `dev1` in application `app1` :

```bash
$ curl --location --header 'Authorization: Bearer NNSXS.XXXXXXXXX' --header 'Accept: application/json' \

>      --header 'Content-Type: application/json' --request POST 'https://thethings.example.com/api/v3/events' \

Reviewer: Here you use quotes around the URL, not above. Please pick one. If you do, then you don't need to escape `?`

     --data-raw '{
    "identifiers":[{
        "device_ids":{
            "device_id":"dev1",
            "application_ids":{"application_id":"app1"}
        }
    }]
}'
```

If you want to create a device, perform multi step actions, or write shell scripts, it's best to use the [CLI]({{< ref "getting-started/cli" >}}).

> If you really, really want to do something like register a device manually, you need to make calls to the Identity Server, Join Server, Network Server and Application Server. 

Reviewer 2: We need to tell people which fields go where. This is already described in the API reference, so please link to `reference/api/end_device`.

> If you really, really want to do something like register a device manually, you need to make calls to the Identity Server, Join Server, Network Server and Application Server. 

Reviewer: ```suggestion
If you want to register a device via the HTTP API, you need to make calls to the Identity Server, Join Server, Network Server and Application Server. 
```

It's highly unlikely that people will use cURL _manually_; these are mostly examples for developers to integrate in their software for automated onboarding. That is a very valid reason to use the HTTP API directly.

> To register a device `newdev1` in application `app1`, first, register the `DevEUI`, `JoinEUI` and cluster addresses in the Identity Server:

Reviewer 2: This is also where you register a friendly name, description, attributes, locations, pictures etc. Again: the exact fields are described in the API reference.

```bash
$ curl --location \

>      --header 'Authorization: Bearer NNSXS.XXXXXXXXX' --header 'Content-Type: application/json' \

Reviewer 2: Let's be consistent and also send the `Accept` header here and below.

     --request POST 'https://thethings.example.com/api/v3/applications/app1/devices' \
     --data-raw '{
      "end_device": {
        "ids": {
          "device_id": "newdev1",
          "dev_eui": "0000000000000000",
          "join_eui": "0000000000000000"
        },
        "join_server_address": "thethings.example.com",
        "network_server_address": "thethings.example.com",
        "application_server_address": "thethings.example.com"
      },
      "field_mask": {
        "paths": [
          "join_server_address",
          "network_server_address",
          "application_server_address",
          "ids.dev_eui",
          "ids.join_eui"
        ]
      }
    }'
```

> Register MAC settings in the Network Server:

Reviewer 2: This doesn't really describe what's happening below, where we're not actually setting the `mac_setting`.

```bash
$ curl --location \
     --header 'Authorization: Bearer NNSXS.XXXXXXXXX' --header 'Content-Type: application/json' \
     --request PUT 'https://thethings.example.com/api/v3/ns/applications/app1/devices/newdev1' \
     --data-raw '{
      "end_device": {
        "supports_join": true,
        "lorawan_version": "1.0.2",
        "ids": {
          "device_id": "newdev1",
          "dev_eui": "0000000000000000",
          "join_eui": "0000000000000000"
        },

>         "supports_class_c": false,

Reviewer 2: Not sure if we should confuse people with Class C in a "Getting Started". Setting it to `true` when not actually true, will probably cause confusion.

Reviewer 3: Yes, no need to set this field at all, just assume class A and use the minimum amount of data required

        "lorawan_phy_version": "1.0.2-b",
        "frequency_plan_id": "EU_863_870_TTN"
      },
      "field_mask": {

>         "paths": [

Reviewer 3: I would prefer only the required fields here

          "supports_class_c",
          "lorawan_phy_version",
          "frequency_plan_id"
        ]
      }
    }'
```

Register the `DevEUI` and `JoinEUI` in the Application Server:

```bash
$ curl --location \
     --header 'Authorization: Bearer NNSXS.XXXXXXXXX' --header 'Content-Type: application/json' \
     --request PUT 'https://thethings.example.com/api/v3/as/applications/app1/devices/newdev1' \
     --data-raw '{
      "end_device": {
        "ids": {
          "device_id": "newdev1",
          "dev_eui": "0000000000000000",
          "join_eui": "0000000000000000"
        }
      },
      "field_mask": {
        "paths": [
          "ids.device_id",
          "ids.dev_eui",
          "ids.join_eui"
        ]
      }
    }'
```

Finally, register the `DevEUI`, `JoinEUI`, cluster addresses, and keys in the Join Server:

```bash
$ curl --location \
     --header 'Authorization: Bearer NNSXS.XXXXXXXXX' --header 'Content-Type: application/json' \
     --request PUT 'https://thethings.example.com/api/v3/js/applications/app1/devices/newdev1' \
     --data-raw '{
      "end_device": {
        "ids": {
          "device_id": "newdev1",
          "dev_eui": "0000000000000000",

>           "join_eui": "0000000000000000"

Reviewer 3: For `key` you use `X`s, here you use `0`s, I'd prefer a consistent format for context-specific data, perhaps just use `X` to fool-proof this, such that people don't start copy-pasting commands and creating a bunch of 0000000000000000 DevEUI and JoinEUI devices?

        },
        "network_server_address": "thethings.example.com",
        "application_server_address": "thethings.example.com",
        "network_server_kek_label": "",
        "application_server_kek_label": "",

>         "application_server_id": "",

Reviewer 2: I'm pretty sure these are optional (correct me if I'm wrong, @rvolosatovs), so you can omit them (and the field paths below) to simplify this call a bit.

Reviewer 3: Indeed, no need to set zero values (so no `null` and no `""`)

        "net_id": null,
        "root_keys": {
          "app_key": {

>             "key": "XXXXXXXXX"

Reviewer 3: perhaps align this with the actual key length?

## HTTP Queries

Additional fields may be specified in HTTP requests by appending them as query string parameters. For example, to request the `name`, `description`, and `locations` of devices in an `EndDeviceRegistry.Get` request, add these fields to the `field_mask` field. To get this data for device `dev1` in application `app1`:

```bash
$ curl --location --header "Authorization: Bearer NNSXS.XXXXXXXXX" https://thethings.example.com/api/v3/applications/app1/devices/dev1?field_mask=name,description,locations
```

The same request in tenant `tenant1` on a multi-tenant deployment:

```bash
$ curl --location --header "Authorization: Bearer NNSXS.XXXXXXXXX" https://tenant1.thethings.example.com/api/v3/applications/app1/devices/dev1?field_mask=name,description,locations
```

Fields may also be specified as a JSON object in a POST request.

To get a stream of events for device `dev1` in application `app1` :

```bash
$ curl --location \
     --header 'Authorization: Bearer NNSXS.XXXXXXXXX' \

>      --header 'Accept: application/json' \

Reviewer 2: Why no `Accept` header in the two requests above?

     --header 'Content-Type: application/json' \
     --request POST \
     'https://thethings.example.com/api/v3/events' \
     --data-raw '{
    "identifiers":[{
        "device_ids":{
            "device_id":"dev1",
            "application_ids":{"application_id":"app1"}
        }
    }]
}'
```

If you want to create a device, perform multi step actions, or write shell scripts, it's best to use the [CLI]({{< ref "getting-started/cli" >}}).

> If you really, really want to do something like register a device manually, you need to make calls to the `Identity Server`, `Join Server`, `Network Server` and `Application Server`. 

Reviewer 2: 1. Refer to the API reference to see which fields belong on which server.
2. We don't write "Identity Server" etc. as `inline code`

Author: Just took them from the console

```bash
$ curl --location \
  --header "Authorization: Bearer NNSXS.XXXXXXXXX" \
  --header 'Accept: application/json' \
  --header 'User-Agent: my-integration / my-integration-version' \
  'https://thethings.example.com/api/v3/applications/app1/devices/dev1?field_mask=name,description,locations'
```

The same request in tenant `tenant1` on a multi-tenant deployment:

```bash
$ curl --location \
  --header "Authorization: Bearer NNSXS.XXXXXXXXX" \
  --header 'Accept: application/json' \
  --header 'User-Agent: my-integration / my-integration-version' \
  'https://tenant1.thethings.example.com/api/v3/applications/app1/devices/dev1?field_mask=name,description,locations'
```

Fields may also be specified as a JSON object in a POST request.

To get a stream of events for device `dev1` in application `app1` :

```bash
$ curl --location \
  --header 'Authorization: Bearer NNSXS.XXXXXXXXX' \

>   --header 'Accept: application/json' \

Reviewer 2: The event stream accept header should be `Accept: text/event-stream`. See also the API reference: https://thethingsstack.io/reference/api/events/

Reviewer 2: And maybe we should document what a `text/event-stream` response looks like. 

Background: https://developer.mozilla.org/en-US/docs/Web/API/Server-sent_events/Using_server-sent_events

  --header 'Content-Type: application/json' \
  --header 'User-Agent: my-integration / my-integration-version' \
  --request POST \
  --data-raw '{
    "identifiers":[{
        "device_ids":{
            "device_id":"dev1",
            "application_ids":{"application_id":"app1"}
        }
    }]
  }' \
  'https://thethings.example.com/api/v3/events'
```

If you want to create a device, perform multi step actions, or write shell scripts, it's best to use the [CLI]({{< ref "getting-started/cli" >}}).

> If you want to do something like register a device manually, you need to make calls to the Identity Server, Join Server, Network Server and Application Server. See the [API Reference]({{< ref "reference/api" >}}) for detailed information about which messages go to which endpoints. 

Reviewer: ```suggestion
If you want to do something like registering a device directly via the API, you need to make calls to the Identity Server, Join Server, Network Server and Application Server. See the [API Reference]({{< ref "reference/api" >}}) for detailed information about which messages go to which endpoints. 
```

        "ids.join_eui"
      ]
    }
  }' \
  'https://thethings.example.com/api/v3/as/applications/app1/devices/newdev1'
```

Finally, register the `DevEUI`, `JoinEUI`, cluster addresses, and keys in the Join Server:

```bash
$ curl --location \
  --header 'Authorization: Bearer NNSXS.XXXXXXXXX' --header 'Content-Type: application/json' \
  --header 'User-Agent: my-integration / my-integration-version' \
  --request PUT \
  --data-raw '{
    "end_device": {
      "ids": {
        "device_id": "newdev1",
        "dev_eui": "XXXXXXXXXXXXXXXX",
        "join_eui": "XXXXXXXXXXXXXXXX"
      },
      "network_server_address": "thethings.example.com",
      "application_server_address": "thethings.example.com",
      "root_keys": {
        "app_key": {
          "key": "XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX"
        }
      }
    },
    "field_mask": {
      "paths": [
        "network_server_address",
        "application_server_address",
        "ids.device_id",
        "ids.dev_eui",
        "ids.join_eui",
        "root_keys.app_key.key"
      ]
    }
  }' \
  'https://thethings.example.com/api/v3/js/applications/app1/devices/newdev1'
```

## Best Practices

### Send a User-Agent Header

> Set a User-Agent header containing your integration name and version. That way, a network operator can target specific integrations and versions in their firewall rules in case some integration misbehaves, instead of blocking all versions of curl or your integration completely.

Reviewer: ```suggestion
Set the `User-Agent` HTTP header containing your integration name and version. That way, a network operator can help finding out potential issues using the logs.
```

`User-Agent` should never be used for security

> ### Look for X-Ratelimit-* Response Headers

Reviewer: ```suggestion
### Respect `X-Ratelimit-*` Response Headers
```

> {{% tts %}} sends responses containing information about how many requests your integration has made and how many are remaining. See [here](https://auth0.com/docs/policies/rate-limit-policy#review-http-response-headers) for more information about how to act on this information.

Reviewer 2: It feels a bit weird to refer to the API documentation of Auth0 for this. 

If we want to provide background, we could point to the internet-draft (https://tools.ietf.org/id/draft-polli-ratelimit-headers-03.html) and explain that we use the `X-` prefix until it becomes an RFC.

> ### Look for X-Warning Headers

Reviewer: ```suggestion
### Mind the `X-Warning` Headers
```

        "device_id": "newdev1",
        "dev_eui": "XXXXXXXXXXXXXXXX",
        "join_eui": "XXXXXXXXXXXXXXXX"
      },
      "lorawan_phy_version": "1.0.2-b",
      "frequency_plan_id": "EU_863_870_TTN"
    },
    "field_mask": {
      "paths": [
        "supports_join",
        "lorawan_version",
        "ids.device_id",
        "ids.dev_eui",
        "ids.join_eui",
        "lorawan_phy_version",
        "frequency_plan_id"
      ]
    }
  }' \
  'https://thethings.example.com/api/v3/ns/applications/app1/devices/newdev1'
```

Register the `DevEUI` and `JoinEUI` in the Application Server:

```bash
$ curl --location \
  --header 'Authorization: Bearer NNSXS.XXXXXXXXX' --header 'Content-Type: application/json' \
  --header 'User-Agent: my-integration/my-integration-version' \
  --request PUT \
  --data-raw '{
    "end_device": {
      "ids": {
        "device_id": "newdev1",
        "dev_eui": "XXXXXXXXXXXXXXXX",
        "join_eui": "XXXXXXXXXXXXXXXX"
      }
    },
    "field_mask": {
      "paths": [
        "ids.device_id",
        "ids.dev_eui",
        "ids.join_eui"
      ]
    }
  }' \
  'https://thethings.example.com/api/v3/as/applications/app1/devices/newdev1'
```

> Finally, register the `DevEUI`, `JoinEUI`, cluster addresses, and keys in the Join Server:

Reviewer 3: I guess it would be nice to mention that this is only relevant for OTAA devices?

```bash
$ curl --location \
  --header "Authorization: Bearer NNSXS.XXXXXXXXX" \
  --header 'Accept: application/json' \
  'https://thethings.example.com/api/v3/applications/app1/devices/dev1?field_mask=name,description,locations'
```

The same request in tenant `tenant1` on a multi-tenant deployment:

```bash
$ curl --location \
  --header "Authorization: Bearer NNSXS.XXXXXXXXX" \
  --header 'Accept: application/json' \
  'https://tenant1.thethings.example.com/api/v3/applications/app1/devices/dev1?field_mask=name,description,locations'
```

Fields may also be specified as a JSON object in a POST request.

To get a stream of events for device `dev1` in application `app1` :

```bash
$ curl --location \
  --header 'Authorization: Bearer NNSXS.XXXXXXXXX' \
  --header 'Accept: application/json' \
  --header 'Content-Type: application/json' \
  --request POST 'https://thethings.example.com/api/v3/events' \

>   --data-raw '{

Reviewer: Does this work; order of flags, url, more flags?

Reviewer: My help says;

```
Usage: curl [options...] <url>
```

Author: it does work but it's semantically wrong, i'll rewrite

If you want to create a device, perform multi step actions, or write shell scripts, it's best to use the [CLI]({{< ref "getting-started/cli" >}}).

> If you want to do something like register a device manually, you need to make calls to the Identity Server, Join Server, Network Server and Application Server. See the [API Reference](({{< ref "reference/api" >}})) for detailed information about which messages go to which endpoints. 

Reviewer: One pair of `()` too many

> To register a device `newdev1` in application `app1`, first, register the `DevEUI`, `JoinEUI` and cluster addresses in the Identity Server. This is also where you register a friendly name, description, attributes, location, and more - see all fields in the [API Reference](({{< ref "reference/api" >}})):

Reviewer: Idem
